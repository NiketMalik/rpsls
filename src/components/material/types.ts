import type { MATERIAL_TYPES } from "constants/material";

export interface MaterialProps {
  type?: MATERIAL_TYPES;
  radius?: number;
  isDisabled?: boolean;
  isHighlited?: boolean;
  onClick?: () => void;
}
