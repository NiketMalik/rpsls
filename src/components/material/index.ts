export { Material } from "./Material";
export { DEFAULT_RADIUS } from "./constants";
export type { MaterialProps } from "./types";
