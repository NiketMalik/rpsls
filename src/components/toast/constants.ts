export enum TOAST_ACCENTS {
  "default" = "default",
  "primary" = "primary",
  "secondary" = "secondary",
  "info" = "info",
  "success" = "success",
  "danger" = "danger",
  "warning" = "warning",
}
