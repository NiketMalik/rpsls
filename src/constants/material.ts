export enum MATERIAL_TYPES {
  ROCK = "ROCK",
  PAPER = "PAPER",
  SCISSOR = "SCISSOR",
  LIZARD = "LIZARD",
  SPOCK = "SPOCK",
}
