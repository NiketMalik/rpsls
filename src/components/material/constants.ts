import { MATERIAL_TYPES } from "constants/material";

import rockIcon from "static/images/material-rock.svg";
import paperIcon from "static/images/material-paper.svg";
import scissorIcon from "static/images/material-scissor.svg";
import lizardIcon from "static/images/material-lizard.svg";
import spockIcon from "static/images/material-spock.svg";

export const MATERIAL_ICONS_MAP = {
  [MATERIAL_TYPES.ROCK]: rockIcon,
  [MATERIAL_TYPES.PAPER]: paperIcon,
  [MATERIAL_TYPES.SCISSOR]: scissorIcon,
  [MATERIAL_TYPES.LIZARD]: lizardIcon,
  [MATERIAL_TYPES.SPOCK]: spockIcon,
};

export const DEFAULT_RADIUS = 60;
