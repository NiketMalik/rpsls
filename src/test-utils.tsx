import { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { THEME } from "constants/theme";

import { StoreProvider } from "context/storeContext";
import { ThemeProvider } from "@emotion/react";

const WithGlobalProviders: FC = ({ children }) => {
  return (
    <StoreProvider>
      <ThemeProvider theme={THEME}>{children}</ThemeProvider>
    </StoreProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithGlobalProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
