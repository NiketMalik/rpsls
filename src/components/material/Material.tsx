import { memo } from "react";

import type { MaterialProps } from "./types";

import { DEFAULT_RADIUS, MATERIAL_ICONS_MAP } from "./constants";

import { MaterialContainer, MaterialImage, MaterialHighlight } from "./styles";

export const Material = memo(
  ({ type, radius, isDisabled, isHighlited, onClick }: MaterialProps) => {
    const materialIconUrl = type && MATERIAL_ICONS_MAP[type];

    return (
      <MaterialContainer
        aria-label={type}
        material={type}
        radius={radius || DEFAULT_RADIUS}
        onClick={!isDisabled ? onClick : undefined}
        data-disabled={isDisabled || !materialIconUrl}
        data-is-highlited={!!isHighlited}
        data-testid="material"
      >
        {materialIconUrl && (
          <MaterialImage
            src={materialIconUrl}
            alt={type}
            data-testid="material-image"
          />
        )}
        <MaterialHighlight data-testid="material-highlight" />
      </MaterialContainer>
    );
  },
);
