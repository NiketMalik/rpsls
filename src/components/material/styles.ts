import styled from "@emotion/styled/macro";

import { MATERIAL_TYPES } from "constants/material";

import { getLinearGradient } from "utils/styles";

interface MaterialContainerProps {
  material?: MATERIAL_TYPES;
  radius: number;
}

export const MaterialHighlight = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;

  &::before,
  &::after {
    content: "";
    position: absolute;
    border-radius: 50%;
    background: ${(props) => props.theme.gradients.background.start};
    transition: transform 0.25s ease-in-out;
    pointer-events: none;
  }

  &::before {
    width: 220%;
    height: 220%;
    top: -60%;
    left: -60%;

    opacity: 0.1;
  }

  &::after {
    width: 160%;
    height: 160%;
    top: -30%;
    left: -30%;

    opacity: 0.15;
  }
`;

export const MaterialContainer = styled.button<MaterialContainerProps>`
  position: relative;
  width: ${(props) => props.radius * 2}px;
  height: ${(props) => props.radius * 2}px;
  min-width: ${(props) => props.radius * 2}px;
  min-height: ${(props) => props.radius * 2}px;

  &::before,
  &::after {
    content: "";
    position: absolute;
    border-radius: 50%;
    z-index: 1;
  }

  &::before {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    ${(props) =>
      getLinearGradient(
        props.theme.gradients.material[props.material || "DEFAULT"],
      )}
    opacity: ${(props) => (props.material ? "1" : "0.5")};
    box-shadow: ${(props) =>
      props.material ? "inset 0 -5px rgba(0, 0, 0, 0.15)" : "none"};
  }

  &::after {
    display: ${(props) => (props.material ? "block" : "none")};
    top: 50%;
    left: 50%;
    width: 80%;
    height: 80%;
    transform: translate(-50%, -50%);

    background: white;
    box-shadow: inset 0 7px rgba(0, 0, 0, 0.15);
  }

  &:not([disabled]):active {
    transform: scale(0.95);
  }

  &[data-is-highlited="true"] {
    margin: 30%;

    ${MaterialHighlight} {
      &::before,
      &::after {
        transform: scale(1);
      }
    }
  }

  &[data-is-highlited="false"] {
    ${MaterialHighlight} {
      &::before,
      &::after {
        transform: scale(0);
      }
    }
  }
`;

export const MaterialImage = styled.img`
  position: absolute;
  top: 50%;
  left: 50%;
  width: 45%;
  height: 45%;
  transform: translate(-50%, -50%);
  z-index: 2;
`;
