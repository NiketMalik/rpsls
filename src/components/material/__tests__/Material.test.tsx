import React from "react";
import { fireEvent, render, screen, waitFor } from "test-utils";
import { MATERIAL_TYPES } from "constants/material";
import { MATERIAL_ICONS_MAP } from "../constants";

import { Material } from "../Material";

describe("Material", () => {
  it("renders without error", () => {
    const { container } = render(<Material />);
    const materialEle = screen.getByTestId("material");
    expect(materialEle).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("renders material types without error", async () => {
    const { container, rerender } = render(<Material />);

    await waitFor(
      async () =>
        await Promise.all(
          Object.values(MATERIAL_TYPES).map((materialType) => {
            rerender(<Material type={materialType} />);

            const srcEle = screen.getByTestId(
              "material-image",
            ) as HTMLImageElement;
            expect(srcEle).toBeInTheDocument();
            expect(srcEle.getAttribute("src")).toBe(
              MATERIAL_ICONS_MAP[materialType],
            );
            expect(container).toMatchSnapshot();
          }),
        ),
    );
  });

  it("renders with highlight", () => {
    const { container } = render(<Material isHighlited={true} />);
    const materialEle = screen.getByTestId("material");
    const highlightEle = screen.getByTestId("material-highlight");
    expect(materialEle).toBeInTheDocument();
    expect(highlightEle).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("renders with given redius", () => {
    const { container } = render(<Material radius={500} />);
    const materialEle = screen.getByTestId("material");
    expect(materialEle).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  describe("click action", () => {
    it("fires callback when not disabled", () => {
      const mockClick = jest.fn();
      render(
        <Material
          type={MATERIAL_TYPES.ROCK}
          onClick={mockClick}
          isDisabled={false}
        />,
      );
      const materialEle = screen.getByTestId("material");

      fireEvent.click(materialEle);
      const srcEle = screen.getByTestId("material-image");
      expect(srcEle).toBeInTheDocument();
      expect(mockClick).toBeCalledTimes(1);
    });

    it("does not fire callback when disabled", () => {
      const mockClick = jest.fn();
      render(
        <Material
          type={MATERIAL_TYPES.ROCK}
          onClick={mockClick}
          isDisabled={true}
        />,
      );
      const materialEle = screen.getByTestId("material");

      fireEvent.click(materialEle);
      const srcEle = screen.getByTestId("material-image");
      expect(srcEle).toBeInTheDocument();
      expect(mockClick).toBeCalledTimes(0);
    });
  });
});
