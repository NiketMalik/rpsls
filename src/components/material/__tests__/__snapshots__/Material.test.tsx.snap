// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Material renders material types without error 1`] = `
<div>
  <button
    aria-label="ROCK"
    class="css-1ui2vtv-MaterialContainer ec4hnvt1"
    data-disabled="false"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <img
      alt="ROCK"
      class="css-stx9by-MaterialImage ec4hnvt0"
      data-testid="material-image"
      src="material-rock.svg"
    />
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders material types without error 2`] = `
<div>
  <button
    aria-label="PAPER"
    class="css-uzsmmi-MaterialContainer ec4hnvt1"
    data-disabled="false"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <img
      alt="PAPER"
      class="css-stx9by-MaterialImage ec4hnvt0"
      data-testid="material-image"
      src="material-paper.svg"
    />
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders material types without error 3`] = `
<div>
  <button
    aria-label="SCISSOR"
    class="css-919jtz-MaterialContainer ec4hnvt1"
    data-disabled="false"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <img
      alt="SCISSOR"
      class="css-stx9by-MaterialImage ec4hnvt0"
      data-testid="material-image"
      src="material-scissor.svg"
    />
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders material types without error 4`] = `
<div>
  <button
    aria-label="LIZARD"
    class="css-9k3u6a-MaterialContainer ec4hnvt1"
    data-disabled="false"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <img
      alt="LIZARD"
      class="css-stx9by-MaterialImage ec4hnvt0"
      data-testid="material-image"
      src="material-lizard.svg"
    />
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders material types without error 5`] = `
<div>
  <button
    aria-label="SPOCK"
    class="css-44liza-MaterialContainer ec4hnvt1"
    data-disabled="false"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <img
      alt="SPOCK"
      class="css-stx9by-MaterialImage ec4hnvt0"
      data-testid="material-image"
      src="material-spock.svg"
    />
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders with given redius 1`] = `
<div>
  <button
    class="css-1c4hhev-MaterialContainer ec4hnvt1"
    data-disabled="true"
    data-is-highlited="false"
    data-testid="material"
    radius="500"
  >
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders with highlight 1`] = `
<div>
  <button
    class="css-14jmf8d-MaterialContainer ec4hnvt1"
    data-disabled="true"
    data-is-highlited="true"
    data-testid="material"
    radius="60"
  >
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;

exports[`Material renders without error 1`] = `
<div>
  <button
    class="css-14jmf8d-MaterialContainer ec4hnvt1"
    data-disabled="true"
    data-is-highlited="false"
    data-testid="material"
    radius="60"
  >
    <div
      class="css-1nc6hb2-MaterialHighlight ec4hnvt2"
      data-testid="material-highlight"
    />
  </button>
</div>
`;
